File: src/main/java/templemore/onx/version2/Line.java
package templemore.onx.version2;

import java.util.ArrayList;
import java.util.List;

/**
 * @author Chris Turner
 */
class Line {

    private final int row;
    private final int column;
    private final Token[] tokens;

    public Line(int row, int column, Token[] tokens) {
        this.row = row;
        this.column = column;
        this.tokens = tokens;
    }

    public Token[] getTokens() {
        return tokens;
    }

    public int emptySpaces() {
        int empty = 0;
        for ( final Token token : tokens ) {
            if ( token == null ) empty++;
        }
        return empty;
    }

    public Position getEmptyPosition() {
        return getEmptyPositions()[0];
    }

    public Position[] getEmptyPositions() {
        final List<Position> result = new ArrayList<Position>();
        for ( int i = 0; i < tokens.length; i++ ) {
            if ( tokens[i] == null ) {
                if ( column == -1 ) result.add(new Position(row, i));
                else if ( row == -1 ) result.add(new Position(i, column));
                else if ( i == 0 ) result.add(new Position(row, column));
                else if ( i == 1 ) result.add(new Position(1, 1));
                else if ( row == 0 ) result.add(new Position(2, 2));
                else result.add(new Position(column, row));
            }
        }
        return result.toArray(new Position[result.size()]);
    }

    public boolean isAllSameToken(final Token token) {
        for ( final Token t : tokens ) {
            if ( (t != null) && (!t.equals(token)) ) return false;
        }
        return true;
    }
}

File: src/main/scala/templemore/onx/version3/Line.scala
package templemore.onx.version3

import Token._

/**
 * @author Chris Turner
 */
case class Line(row: Int, column: Int, tokens: Array[Token]) {

  def emptySpaces = {
    var empty = 0
    tokens.foreach { token => if ( token == null ) empty += 1 }
    empty
  }

  def emptyPosition = emptyPositions()(0)

  def emptyPositions(): Array[Position] = {
    var result = List[Position]()
    for ( i <- 0 to (tokens.length - 1) ) {
      if ( tokens(i) == null ) {
        if ( column == -1 ) result = Position(row, i) :: result
        else if ( row == -1 ) result = Position(i, column) :: result
        else if ( i == 0 ) result = Position(row, column) :: result
        else if ( i == 1 ) result = Position(1, 1) :: result
        else if ( row == 0 ) result = Position(2, 2) :: result
        else result = Position(column, row) :: result
      }
    }
    result.toArray
  }

  def allSameToken_?(token: Token) = tokens.forall { t => !((t != null) && (t != token)) }
}
